OpenSim/Common/ComponentPath.h

```diff
@@ -0,0 +1,106 @@
+ *                                                                            *
+ * Copyright (c) 2005-2016 Stanford University and the Authors                *
+ * Author(s): Carmichael Ong                                                  *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include "Path.h"
+
+namespace OpenSim {
+
+//==============================================================================
+//                            OPENSIM COMPONENT PATH
+//==============================================================================
+/**
+* A class for handling Paths for Components, deriving from Path. A ComponentPath
+* uses a forward-slash ('/') as a separator always. It also specifies invalid
+* characters in a Component name:
+* - back-slash ('\\')
+* - forward-slash ('/')
+* - asterisk ('*')
+* - plus-sign ('+')
+*
+* @author Carmichael Ong
+*/
+
+class OSIMCOMMON_API ComponentPath : public Path {
+
+public:
+    // Constructors
+    /// Default Constructor
+    ComponentPath() = default;
+    
+    /// Construct a ComponentPath from a string. This will clean up the
+    /// path, removing and resolving "." and ".." when possible.
+    ComponentPath(const std::string path);
+
+    /// Constructor a ComponentPath from a vector that contains all subtree
+    /// node names and a bool that indicates if the path is an absolute path.
+    ComponentPath(std::vector<std::string> pathVec, bool isAbsolute);
+
+    // Operators
+    bool operator==(const ComponentPath& other) const
+    {
+        return this->getString() == other.getString();
+    }
+
+    bool operator!=(const ComponentPath& other) const
+    {
+        return this->getString() != other.getString();
+    }
+
+    // Override virtual functions
+    const char getSeparator() const override { return separator; };
+    const std::string getInvalidChars() const override { return invalidChars; };
+
+    /// Get an absolute path by resolving it relative to a given otherPath.
+    /// If the current Path is already absolute, return the same Path.
+    ComponentPath formAbsolutePath(ComponentPath* otherPath);
```

Comment:
Should take `const ComponentPath& otherPath`, and should also be a const function.